In [26]:
import numpy as np
import pandas as pd
from pathlib import Path
import matplotlib.pyplot as plt

In [2]:
train_df = pd.read_csv(Path('Resources/2019loans.csv'))
test_df = pd.read_csv(Path('Resources/2020Q1loans.csv'))

In [3]:
train_df

Unnamed: 0   index  loan_amnt  int_rate  installment home_ownership  \
0           57107   57107    13375.0    0.1797       483.34       MORTGAGE   
1          141451  141451    21000.0    0.1308       478.68       MORTGAGE   
2          321143  321143    20000.0    0.1240       448.95       MORTGAGE   
3           11778   11778     3000.0    0.1240       100.22           RENT   
4          169382  169382    30000.0    0.1612      1056.49       MORTGAGE   
...           ...     ...        ...       ...          ...            ...   
12175      354912  354912    19975.0    0.2565       801.09           RENT   
12176      354944  354944    15000.0    0.1774       540.34           RENT   
12177      354973  354973     3600.0    0.1862       131.28           RENT   
12178      355002  355002    15000.0    0.0881       475.68       MORTGAGE   
12179      355307  355307    15000.0    0.1774       540.34           RENT   

       annual_inc verification_status loan_status pymnt_plan  ...  \
0        223000.0        Not Verified    low_risk          n  ...   
1        123000.0     Source Verified    low_risk          n  ...   
2        197000.0     Source Verified    low_risk          n  ...   
3         45000.0        Not Verified    low_risk          n  ...   
4        133000.0     Source Verified    low_risk          n  ...   
...           ...                 ...         ...        ...  ...   
12175     28000.0        Not Verified   high_risk          n  ...   
12176     50000.0            Verified   high_risk          n  ...   
12177     60000.0        Not Verified   high_risk          n  ...   
12178     62000.0     Source Verified   high_risk          n  ...   
12179     54080.0        Not Verified   high_risk          n  ...   

       pct_tl_nvr_dlq  percent_bc_gt_75  pub_rec_bankruptcies  tax_liens  \
0               100.0              50.0                   0.0        0.0   
1                85.0              33.3                   0.0        0.0   
2                85.7              33.3                   0.0        0.0   
3               100.0              16.7                   1.0        0.0   
4               100.0              66.7                   0.0        0.0   
...               ...               ...                   ...        ...   
12175           100.0              16.7                   0.0        0.0   
12176            90.5              11.1                   0.0        0.0   
12177           100.0               0.0                   0.0        0.0   
12178           100.0               0.0                   0.0        0.0   
12179            90.0             100.0                   0.0        0.0   

       tot_hi_cred_lim  total_bal_ex_mort  total_bc_limit  \
0             577150.0           122018.0         32000.0   
1             132750.0            27896.0         15900.0   
2             628160.0           114043.0         22600.0   
3              42006.0            20761.0         19900.0   
4             283248.0           109056.0         79500.0   
...                ...                ...             ...   
12175          50055.0            28192.0         18700.0   
12176          70324.0            57025.0         13300.0   
12177          83765.0            55156.0         14800.0   
12178         189930.0            23748.0          7000.0   
12179          48968.0            29786.0          2900.0   

      total_il_high_credit_limit  hardship_flag  debt_settlement_flag  
0                       170200.0              N                     N  
1                        35398.0              N                     N  
2                        90340.0              N                     N  
3                        15406.0              N                     N  
4                        58778.0              N                     N  
...                          ...            ...                   ...  
12175                    19055.0              N                     N  
12176              

In [4]:
# Convert categorical data to numeric and separate target feature for training data
train_df = pd.get_dummies(train_df)
train_col = train_df.columns

In [5]:
# Convert categorical data to numeric and separate target feature for testing data
test_df = pd.get_dummies(test_df)
test_col = test_df.columns

In [6]:
# add missing dummy variables to testing set
missing_variables = []
for col in train_col:
    if col not in test_col:
        missing_variables.append(col)

print(missing_variables)

['debt_settlement_flag_Y']


In [7]:
test_df.astype('int')
train_df.astype('int')

Unnamed: 0   index  loan_amnt  int_rate  installment  annual_inc  dti  \
0           57107   57107      13375         0          483      223000   29   
1          141451  141451      21000         0          478      123000   11   
2          321143  321143      20000         0          448      197000   11   
3           11778   11778       3000         0          100       45000   18   
4          169382  169382      30000         0         1056      133000   27   
...           ...     ...        ...       ...          ...         ...  ...   
12175      354912  354912      19975         0          801       28000   28   
12176      354944  354944      15000         0          540       50000   23   
12177      354973  354973       3600         0          131       60000   28   
12178      355002  355002      15000         0          475       62000   11   
12179      355307  355307      15000         0          540       54080   21   

       delinq_2yrs  inq_last_6mths  open_acc  ...  loan_status_low_risk  \
0                0               0        15  ...                     1   
1                2               0        16  ...                     1   
2                0               0        12  ...                     1   
3                0               0        12  ...                     1   
4                0               2        13  ...                     1   
...            ...             ...       ...  ...                   ...   
12175            0               0        15  ...                     0   
12176            4               0        16  ...                     0   
12177            0               1        14  ...                     0   
12178            0               0         5  ...                     0   
12179            1               1         5  ...                     0   

       pymnt_plan_n  initial_list_status_f  initial_list_status_w  \
0                 1                      0                      1   
1                 1                      0                      1   
2                 1                      0                      1   
3                 1                      0                      1   
4                 1                      0                      1   
...             ...                    ...                    ...   
12175             1                      0                      1   
12176             1                      0                      1   
12177             1                      0                      1   
12178             1                      0                      1   
12179             1                      1                      0   

       application_type_Individual  application_type_Joint App  \
0                                1                           0   
1                                1                           0   
2                                1                           0   
3                                1                           0   
4                                1                           0   
...                            ...                         ...   
12175                            1                           0   
12176                            1                           0   
12177                            1                           0   
12178                            0                           1   
12179                            1                           0   

       hardship_flag_N  hardship_flag_Y  debt_settlement_flag_N  \
0                    1                0                       1   
1                    1                0                       1   
2                    1                0                       1   
3                    1                0                       1   
4                    1                0                       1   
...                ...              ...                     ...   
12175                1                0               

In [32]:
test_df.columns

Index(['Unnamed: 0', 'index', 'loan_amnt', 'int_rate', 'installment',
       'annual_inc', 'dti', 'delinq_2yrs', 'inq_last_6mths', 'open_acc',
       'pub_rec', 'revol_bal', 'total_acc', 'out_prncp', 'out_prncp_inv',
       'total_pymnt', 'total_pymnt_inv', 'total_rec_prncp', 'total_rec_int',
       'total_rec_late_fee', 'recoveries', 'collection_recovery_fee',
       'last_pymnt_amnt', 'collections_12_mths_ex_med', 'policy_code',
       'acc_now_delinq', 'tot_coll_amt', 'tot_cur_bal', 'open_acc_6m',
       'open_act_il', 'open_il_12m', 'open_il_24m', 'mths_since_rcnt_il',
       'total_bal_il', 'il_util', 'open_rv_12m', 'open_rv_24m', 'max_bal_bc',
       'all_util', 'total_rev_hi_lim', 'inq_fi', 'total_cu_tl', 'inq_last_12m',
       'acc_open_past_24mths', 'avg_cur_bal', 'bc_open_to_buy', 'bc_util',
       'chargeoff_within_12_mths', 'delinq_amnt', 'mo_sin_old_il_acct',
       'mo_sin_old_rev_tl_op', 'mo_sin_rcnt_rev_tl_op', 'mo_sin_rcnt_tl',
       'mort_acc', 'mths_since_recent_bc'

In [8]:
test_df['debt_settlement_flag_Y'] = abs(test_df['debt_settlement_flag_N']-1)

In [17]:
# Train the Logistic Regression model on the unscaled data and print the model score
from sklearn.linear_model import LogisticRegression
classifier = LogisticRegression(max_iter=10000000)
classifier

LogisticRegression(max_iter=10000000)

In [36]:
X_train= train_df.drop(['loan_status_low_risk'], axis=1).drop(['loan_status_high_risk'], axis=1)
y_train= train_df['loan_status_low_risk']
X_test= test_df.drop(['loan_status_low_risk'], axis=1).drop(['loan_status_high_risk'], axis=1)
y_test= test_df['loan_status_low_risk']

In [37]:
classifier.fit(X_train, y_train)

LogisticRegression(max_iter=10000000)

In [38]:
print(f"Training Data Score: {classifier.score(X_train, y_train)}")
print(f"Testing Data Score: {classifier.score(X_test, y_test)}")

Training Data Score: 0.7008210180623974
Testing Data Score: 0.5648660144619311


In [40]:
# Train a Random Forest Classifier model and print the model score
from sklearn.ensemble import RandomForestClassifier
clf = RandomForestClassifier(random_state=1, n_estimators=500).fit(X_train, y_train)
print(f'Training Score: {clf.score(X_train, y_train)}')
print(f'Testing Score: {clf.score(X_test, y_test)}')

Training Score: 1.0
Testing Score: 0.6180348787749894


In [41]:
# Scale the data
from sklearn.preprocessing import StandardScaler
scaler = StandardScaler().fit(X_train)
X_train_scaled = scaler.transform(X_train)
X_test_scaled = scaler.transform(X_test)


In [42]:
# Train the Logistic Regression model on the scaled data and print the model score
classifier.fit(X_train_scaled, y_train)
print(f"Training Data Score: {classifier.score(X_train, y_train)}")
print(f"Testing Data Score: {classifier.score(X_test, y_test)}")

Training Data Score: 0.5110016420361247
Testing Data Score: 0.5491280306252658


In [43]:
# Train a Random Forest Classifier model on the scaled data and print the model score
clf = RandomForestClassifier(random_state=1, n_estimators=500).fit(X_train_scaled, y_train)
print(f'Training Score: {clf.score(X_train_scaled, y_train)}')
print(f'Testing Score: {clf.score(X_test_scaled, y_test)}')

Training Score: 1.0
Testing Score: 0.6193109315185028


### Conclusions

Overall, I felt that my expectations were mostly confirmed, as the logistic regression failed to classify, even when scaled, very far above a coin toss. The randomForest was better as I had thought it would be, but I'm surprised it only was able to yeild a 62% hit rate. This forest model was clearly overfitting on the train set, and perhaps addressing this would increase testing accuracy.